### PR TITLE
`ls`: skip reading metadata

### PR DESCRIPTION
**Pull request by tertsdiepraam**
I have drastically reduced the number of `statx` syscalls in `ls`. This is achieved with the [`once_cell`](https://crates.io/crates/once_cell) crate, which retrieves the metadata only when needed. Furthermore, the `DirEntry` struct provides a cheap way to get the file type, so I added that to the `PathData` struct in a similar way to the metadata. This ensures that for most files, no `statx` syscalls are needed. For example, on the Firefox source code, the number of `statx` calls went from 300,000 to 4.

Finally, I've removed the `lossy_string` attribute of `PathData` and replaced it with `Path::display()` calls to reduce the number of calls of `to_string_lossy`.

While experimenting, I found that changing `to_lowercase` to `to_ascii_lowercase` in this line increases performance by about a factor of 2:
```rust
Sort::Name => entries.sort_by_key(|k| k.file_name.to_lowercase()),
```
This suggests that we should look into either writing or finding a better way to do a case-insensitive sort on the filenames.

On the syscall side of performance, we could look into reducing `write` calls to further improve performance a bit.

# Before
```
Benchmark #1: ls -R
  Time (mean ± σ):     518.6 ms ±   1.5 ms    [User: 365.1 ms, System: 150.7 ms]
  Range (min … max):   516.7 ms … 521.4 ms    10 runs
 
Benchmark #2: ../coreutils/target/release/coreutils ls -R
  Time (mean ± σ):      2.622 s ±  0.011 s    [User: 2.131 s, System: 0.486 s]
  Range (min … max):    2.602 s …  2.642 s    10 runs
 
Summary
  'ls -R' ran
    5.06 ± 0.03 times faster than '../coreutils/target/release/coreutils ls -R'
```

```
Benchmark #1: ls -aR
  Time (mean ± σ):      1.695 s ±  0.005 s    [User: 1.169 s, System: 0.517 s]
  Range (min … max):    1.686 s …  1.702 s    10 runs
 
Benchmark #2: ../coreutils/target/release/coreutils ls -aR
  Time (mean ± σ):      8.819 s ±  0.025 s    [User: 7.080 s, System: 1.721 s]
  Range (min … max):    8.788 s …  8.864 s    10 runs
 
Summary
  'ls -aR' ran
    5.20 ± 0.02 times faster than '../coreutils/target/release/coreutils ls -aR'
```

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ------------------
 46,32    0,697715           2    306690         1 statx
 35,28    0,531408           1    344064           write
  9,32    0,140390           3     37436           getdents64
  3,43    0,051715           2     18697           openat
  3,06    0,046070           2     18697           close
  2,58    0,038857           2     18697           newfstatat
  0,01    0,000159          22         7           munmap
  0,00    0,000075           2        29           brk
  0,00    0,000022           0        31           mmap
  0,00    0,000008           8         1           mremap
  0,00    0,000002           0         3           sigaltstack
  0,00    0,000000           0         9           read
  0,00    0,000000           0         1           poll
  0,00    0,000000           0         9           mprotect
  0,00    0,000000           0         7           rt_sigaction
```

# After
```
Benchmark #1: ls -R
  Time (mean ± σ):     519.1 ms ±   0.9 ms    [User: 355.1 ms, System: 161.0 ms]
  Range (min … max):   517.6 ms … 520.4 ms    10 runs
 
Benchmark #2: ../coreutils/target/release/coreutils ls -R
  Time (mean ± σ):      2.213 s ±  0.006 s    [User: 1.977 s, System: 0.232 s]
  Range (min … max):    2.202 s …  2.221 s    10 runs
 
Summary
  'ls -R' ran
    4.26 ± 0.01 times faster than '../coreutils/target/release/coreutils ls -R'
```
```
Benchmark #1: ls -aR
  Time (mean ± σ):      1.922 s ±  0.010 s    [User: 1.350 s, System: 0.565 s]
  Range (min … max):    1.906 s …  1.938 s    10 runs
 
Benchmark #2: ../coreutils/target/release/coreutils ls -aR
  Time (mean ± σ):      8.503 s ±  0.310 s    [User: 7.648 s, System: 0.844 s]
  Range (min … max):    8.188 s …  9.138 s    10 runs
 
Summary
  'ls -aR' ran
    4.42 ± 0.16 times faster than '../coreutils/target/release/coreutils ls -aR'
```

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ------------------
 66,11    0,449730           1    344064           write
 16,99    0,115597           3     37436           getdents64
  6,53    0,044392           2     18697           openat
  5,51    0,037481           2     18697           close
  4,82    0,032785           1     18697           newfstatat
  0,02    0,000134           2        56           brk
  0,02    0,000131          18         7           munmap
  0,00    0,000012          12         1           mremap
  0,00    0,000002           0        31           mmap
  0,00    0,000000           0         9           read
  0,00    0,000000           0         1           poll
  0,00    0,000000           0         9           mprotect
  0,00    0,000000           0         7           rt_sigaction
  0,00    0,000000           0         1           rt_sigprocmask
  0,00    0,000000           0         1         1 ioctl
  0,00    0,000000           0         6           pread64
  0,00    0,000000           0         1         1 access
  0,00    0,000000           0         1           execve
  0,00    0,000000           0         3           sigaltstack
  0,00    0,000000           0         2         1 arch_prctl
  0,00    0,000000           0         1           futex
  0,00    0,000000           0         1           sched_getaffinity
  0,00    0,000000           0         1           set_tid_address
  0,00    0,000000           0         1           set_robust_list
  0,00    0,000000           0         2           prlimit64
  0,00    0,000000           0         1           getrandom
  0,00    0,000000           0         4         1 statx
------ ----------- ----------- --------- --------- ------------------
100,00    0,680264           1    437738         4 total
```

### CODE REVIEW

File: src/uu/ls/src/ls.rs [hunk 1]
```diff
@@ -1145,32 +1185,28 @@ fn sort_entries(entries: &mut Vec<PathData>, config: &Config) {
 fn is_hidden(file_path: &DirEntry) -> bool {
     let metadata = fs::metadata(file_path.path()).unwrap();
     let attr = metadata.file_attributes();
-    ((attr & 0x2) > 0) || file_path.file_name().to_string_lossy().starts_with('.')
-}
-
-#[cfg(unix)]
-fn is_hidden(file_path: &DirEntry) -> bool {
-    file_path.file_name().to_string_lossy().starts_with('.')
+    ((attr & 0x2) > 0)
```
Comment (sylvestre) [inline src/uu/ls/src/ls.rs:1188]:
```suggestion
    (attr & 0x2) > 0
```

**Comment by sylvestre** [comment]
Looks great :)

**Comment by sylvestre** [comment]
it is now conflicting (sorry), could you please have a look?
thanks

**Comment by tertsdiepraam** [comment]
Makes sense with the other PR's getting merged! I'll take a look!

**Comment by sylvestre** [comment]
Thanks. I tried but I will make mistakes :)

**Pull request merged**